cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -191,4 +195,48 @@ public void stopProgram(Id.Program programId, RunId runId) throws ExecutionExcep
     }
     return null;
   }
+
+  /**
+   * Check against the {@link ProgramRuntimeService} to see if the program is running and update the run record
+   * accordingly.
+   *
+   * @param runRecord The target run record to validate.
+   * @param programId The program id for this run record.
+   * @return validated {@link RunRecord} instance
+   */
+  public RunRecord validateRunRecord(final RunRecord runRecord, Id.Program programId, ProgramType programType)
+    throws ExecutionException, InterruptedException {
+    ProgramRuntimeService.RuntimeInfo runtimeInfo = findRuntimeInfo(programId, programType);
+    // check if it is not running  but the run record says it is update the run record for that program
+    ProgramRunStatus status = runRecord.getStatus();
+    if (status == ProgramRunStatus.RUNNING && runtimeInfo == null) {
+      RunId runId = RunIds.fromString(runRecord.getPid());
+      store.setStop(programId, runId.getId(), System.currentTimeMillis(),
```

Comment:
To do this properly, you need to do a checkAndPut-like operation. Or simply, within a transaction, read the run status in the run record, and if it is still running, update it to failed; or if it is not running any more, then its status has update since you queried the first time, and the correct status is now already in the run record.